les/api.go

```diff
@@ -17,16 +17,400 @@
+// of the next potential result is included in the map with an empty structure
+// assigned to it.
+func (api *PrivateLightServerAPI) PriorityClientInfo(start, stop enode.ID, maxCount int) map[enode.ID]map[string]interface{} {
+	res := make(map[enode.ID]map[string]interface{})
+	ids := api.server.clientPool.getPosBalanceIDs(start, stop, maxCount+1)
+	if len(ids) > maxCount {
+		res[ids[maxCount]] = make(map[string]interface{})
+		ids = ids[:maxCount]
+	}
+	api.server.clientPool.forClients(ids, func(client *clientInfo, id enode.ID) {
+		res[id] = api.clientInfo(client, id)
+	})
+	return res
+}
+
+// clientInfo creates a client info data structure
+func (api *PrivateLightServerAPI) clientInfo(c *clientInfo, id enode.ID) map[string]interface{} {
+	info := make(map[string]interface{})
+	if c != nil {
+		info["isConnected"] = true
+		info["capacity"] = c.capacity
+		info["pricing/balance"], info["pricing/negBalance"] = c.balanceTracker.getBalance(mclock.Now())
+		info["pricing/balanceMeta"] = c.balanceMetaInfo
+	} else {
+		info["isConnected"] = false
+		pb := api.server.clientPool.getPosBalance(id)
+		info["pricing/balance"], info["pricing/balanceMeta"] = pb.value, pb.meta
+	}
+	return info
+}
+
+// sendEvent sends an event to the subscribers interested in it. For global events client == nil.
+func (api *PrivateLightServerAPI) sendEvent(clientEvent string, client *clientInfo) {
+	if len(api.subs) == 0 {
+		return
+	}
+
+	event := make(map[string]interface{})
+	event["totalCapacity"], event["totalConnectedCapacity"], event["priorityConnectedCapacity"] = api.server.clientPool.capacityInfo()
+	if client != nil {
+		event["clientEvent"] = clientEvent
+		event["clientId"] = client.id
+		event["clientInfo"] = api.clientInfo(client, client.id)
+	}
+
+	for sub := range api.subs {
+		select {
+		case <-sub.rpcSub.Err():
+			delete(api.subs, sub)
+		case <-sub.notifier.Closed():
+			delete(api.subs, sub)
+		default:
+			sub.notifier.Notify(sub.rpcSub.ID, event)
+		}
+	}
+}
+
+// setParams either sets the given parameters for a single client (if ID is specified)
+// or the default parameters applicable to clients connected in the future
+func (api *PrivateLightServerAPI) setParams(params map[string]interface{}, client *clientInfo, id enode.ID, posFactors, negFactors *priceFactors) (updateFactors bool, err error) {
+	if client != nil {
+		posFactors, negFactors = &client.posFactors, &client.negFactors
+	}
+	defParams := id == enode.ID{}
+loop:
+	for name, value := range params {
+		errValue := func() error {
+			return fmt.Errorf("invalid value for parameter '%s'", name)
+		}
+		setFactor := func(v *float64) {
+			if posFactors != nil {
+				if val, ok := value.(float64); ok && val >= 0 {
+					*v = val / float64(time.Second)
+					updateFactors = true
+				} else {
+					err = errValue()
+				}
+			} else {
+				err = errClientNotConnected
+			}
+		}
+
+		processed := true
+		switch name {
+		case "pricing/timeFactor":
```

Comment:
TBH, is it might too technical deep for operator to set these configurations?

I guess all we need to do is offer the unit of resource client uses, then all the operator needed is just a price per unit.

Reply:
This is something I would like to leave open for experimentation. Usually a script is going to set these parameters and it is pretty easy to hardcode the ratio there.